src/singledocparser.cpp

```diff
@@ -46,6 +46,9 @@ void SingleDocParser::HandleDocument(EventHandler& eventHandler) {
 }
 
 void SingleDocParser::HandleNode(EventHandler& eventHandler) {
+  if (depth > depth_limit) {
+    throw ParserException(m_scanner.mark(), ErrorMsg::BAD_FILE);
```

Comment:
> But I thought it worth mentioning; if you have deeply nested types, then YAML is probably not the right data structure to use.

That's besides the point. This is not about the regular, well designed use-case. This is about user-supplied data, specifically crafted to crash the target... So I think it's acceptable to throw in this case.

> RAII (and, lifetime guarantees) is wonderful :)

Yeah, I was wondering how to do this in C++ again. I got used to the `defer` statement in golang. :)

Honestly, I find the types you suggested to be a little on the heavy side. Isn't there something in the standard library or boost that could cover for this without having to reinvent the wheel?

I like the original implementation I provided because it's just a single integer to keep track of, without that data structure around..